### PR TITLE
SSL Support

### PR DESCRIPTION
**Pull request by AdamMagaluk**
Implements #20

Provides optional ssl support on a per host level. SSL options are specified in hosts option as a `ssl` parameter. `certificate`, `certificateKey`, and `clientCertificate` can be specified using the secret ref syntax. Note if the secret or the secret field is missing the host is ignored.

To enable ssl support in dispatcher you must specify `SSL_CERT` and `SSL_KEY` as env parameters to dispatcher with paths to a default certificate and key that nginx will use for the default ssl host when the hostname on incoming requests does not match any namespace host. This allows nginx to still return a `444`.

Dispatcher parameters added:
* `SSL_PORT`: This is the SSL port nginx will listen on for ssl enabled hosts. _(Default: `443`)_
* `SSL_CERT`: Path to a ssl certificate used for the default ssl nginx server. Enables SSL support.
* `SSL_KEY`: Path to a ssl certificate key used for the default ssl nginx server.
* `SSL_CERT_DIR`: Path to write namespace host's certificates to. _(Default: `/etc/nginx/ssl`)_

A namespace deployment to enable ssl would look like:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: null
  name: test-namespace
  annotations:
    github.com/30x.dispatcher.hosts: |
      {
        "secure.dev": {
          "ssl": {
            "certificate": { "valueFrom": {"secretKeyRef" : {"key":"ssl.crt"}}},
            "certificateKey": { "valueFrom": {"secretKeyRef" : {"key":"ssl.key"}}}
          }
        }
      }
  labels:
    github.com/30x.dispatcher.routable: "true"
    github.com/30x.dispatcher.org: "example"
    github.com/30x.dispatcher.env: "test"

---

apiVersion: v1
kind: Secret
metadata:
  name: routing
  namespace: test-namespace
data:
  ssl.crt: <base64 of cert>
  ssl.key: <base64 of key>
---
```

Nginx output: 

```


server {
  listen 443 ssl;
  server_name secure.dev;
  # SSL Options
  ssl_ciphers HIGH:!aNULL:!MD5:!DH+3DES:!kEDH;
  ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
  ssl_certificate /etc/nginx/ssl/secure.dev/certificate.crt;
  ssl_certificate_key /etc/nginx/ssl/secure.dev/certificate.key;
 
  ...
}
 
# Default server that will just close the connection as if there was no server available
server {
  listen 443 default_server ssl;
  # SSL Options
  ssl_ciphers HIGH:!aNULL:!MD5:!DH+3DES:!kEDH;
  ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
  ssl_certificate ssl/default-server.crt;
  ssl_certificate_key ssl/default-server.key;
 
  location = /dispatcher/status {
    return 200;
  }
 
  location / {
    return 444;
  }
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -93,6 +100,10 @@ Nginx Configuration:
 * `NGINX_MAX_CLIENT_BODY_SIZE`: Configures the max client request body size of nginx. _(Default: `0`, Disables checking of client request body size.)_
 * `NGINX_STATUS_PATH`: Change the nginx status path that returns 200 on the default server. (Default: `/dispatcher/status`)
 * `PORT`: This is the port that nginx will listen on _(Default: `80`)_
+* `SSL_PORT`: This is the SSL port nginx will listen on for ssl enabled hosts. _(Default: `443`)_
+* `SSL_CERT`: Path to a ssl certificate used for the default ssl nginx server. Enables SSL support.
+* `SSL_KEY`: Path to a ssl certificate key used for the default ssl nginx server.
+* `SSL_CERT_DIR`: Path to write namespace host's certificates to. _(Default: `/etc/nginx/ssl`)_
```
Comment (whitlockjc) [inline README.md]: Why is this necessary?  If we own the container, can't we just put them where we want using whatever convention we want?  We don't let the user dictate where the logs are written for example.

File: main.go [hunk 2]
```diff
@@ -102,6 +102,7 @@ func main() {
 
 	// Don't write nginx conf when not in cluster
 	nginx.RunInMockMode = !(kube.RunningInCluster())
+	config.Nginx.RunInMockMode = nginx.RunInMockMode
```
Comment (whitlockjc) [inline main.go]: Maybe we should get rid of the approch on L104 and always require/use the `config` any place where `RunInMockMode` is checked?  _(Minor nit, should not keep this PR from being merged.)_

File: README.md (hunk 1 above)
Comment (AdamMagaluk) [inline README.md]: Are you referring to `SSL_CERT_DIR` specifically or all SSL_* config options. (GH makes it hard to tell sometimes.)

In our deployment we own the Docker container but other people using it may have different needs i guess. I added it initially so i could test easily without running dispatcher as root and having access to `/etc/nginx`. I don't feel to opinionated on the matter though so i could remove it.

File: main.go (hunk 2 above)
Comment (AdamMagaluk) [inline main.go]: Yeah that's probably the right way to do it.

File: README.md (hunk 1 above)
Comment (whitlockjc) [inline README.md]: It was about the `SSL_CERT_DIR`.  I was just wondering why we're letting them choose where in the container, or on disk, these files are stored.

Comment (AdamMagaluk) [inline README.md]: Again i'm not too opinionated, so let me know if you want it removed. Personally i think it's useful to be able to specify that dir, i'd probably also want where the nginx config as well. Would make running locally much easier. Though i can see this being a rabbit hole as well since there are so many nginx knobs we could be tweaking.

**Comment by whitlockjc** [comment]
:shipit:
